sdk/cosmos/azure-cosmos-spark_3/src/main/scala/com/azure/cosmos/spark/TransactionalBulkWriter.scala

```diff
@@ -0,0 +1,1083 @@
+                               clearPendingRetryAction: () => Boolean,
+                               partitionKey: PartitionKey,
+                               objectNode: ObjectNode,
+                               operationContext: OperationContext,
+                               statusCode: Int): Unit = {
+      if (trackPendingRetryAction()) {
+        this.pendingRetries.incrementAndGet()
+      }
+      // this is to ensure the submission will happen on a different thread in background
+      // and doesn't block the active thread
+      val deferredRetryMono = SMono.defer(() => {
+          scheduleWriteInternal(
+              partitionKey,
+              objectNode,
+              new OperationContext(
+                  operationContext.itemId,
+                  operationContext.partitionKeyValue,
+                  operationContext.eTag,
+                  operationContext.attemptNumber + 1,
+                  operationContext.sequenceNumber))
+          if (clearPendingRetryAction()) {
+            this.pendingRetries.decrementAndGet()
+          }
+          SMono.empty
+      })
+
+      if (Exceptions.isTimeout(statusCode)) {
+          deferredRetryMono
+              .delaySubscription(
+                  Duration(
+                      TransactionalBulkWriter.minDelayOn408RequestTimeoutInMs +
+                          scala.util.Random.nextInt(
+                              TransactionalBulkWriter.maxDelayOn408RequestTimeoutInMs - TransactionalBulkWriter.minDelayOn408RequestTimeoutInMs),
+                      TimeUnit.MILLISECONDS),
+                  Schedulers.boundedElastic())
+              .subscribeOn(Schedulers.boundedElastic())
+              .subscribe()
+
+      } else {
+          deferredRetryMono
+              .subscribeOn(Schedulers.boundedElastic())
+              .subscribe()
+      }
+  }
+
+  private val subscriptionDisposable: Disposable = {
+    log.logTrace(s"subscriptionDisposable, Context: ${operationContext.toString} $getThreadInfo")
+
+    val inputFlux = bulkInputEmitter
+      .asFlux()
+      .onBackpressureBuffer()
+      .publishOn(bulkWriterInputBoundedElastic)
+      .doOnError(t => {
+        log.logError(s"Input publishing flux failed, Context: ${operationContext.toString} $getThreadInfo", t)
+      })
+
+    // Use TransactionalBulkExecutor which internally uses SinglePartitionKeyServerBatchRequest
+    // with setAtomicBatch(true) and setShouldContinueOnError(false) for true transactional semantics
+    val cosmosBulkExecutionOptionsImpl = ImplementationBridgeHelpers.CosmosBulkExecutionOptionsHelper
+      .getCosmosBulkExecutionOptionsAccessor
+      .getImpl(cosmosBulkExecutionOptions)
+    
+    val transactionalExecutor = new TransactionalBulkExecutor[Object](
+      container,
+      inputFlux,
+      cosmosBulkExecutionOptionsImpl)
+
+    val bulkOperationResponseFlux: SFlux[CosmosBulkOperationResponse[Object]] =
+      transactionalExecutor
+          .execute()
```

Comment:
I think the retry decision here should not be based on each item[CosmosBulkOperationResponse], but should be based on the whole batch[CosmosBatchResponse], else it may cause wrong grouping of the items during retry

for example:
    item1, item2, item3 belong to batch 1
    item4, item5, item6 belong to batch 2

schedule write for item1
schedule write for item2
schedule write for item3
flush batch 1
schedule write for item4
getting response back from batch 1 and decides to retry
schedule write for item1
---> this will cause flush batch with only item4
schedule write for item5
---> this will cause flush batch with only item1
....


Batch 1: item1, item2, item3